pkg/service/replay/replay.go

```diff
@@ -931,37 +944,62 @@ func (r *Replayer) GetMocks(ctx context.Context, testSetID string, afterTime tim
 	return filtered, unfiltered, err
 }
 
-func (r *Replayer) SetupOrUpdateMocks(ctx context.Context, appID uint64, testSetID string, afterTime, beforeTime time.Time, action MockAction, outgoingOpts models.OutgoingOptions) error {
-
-	if !r.instrument {
-		r.logger.Debug("Keploy will not setup or update the mocks when base path is provided", zap.Any("base path", r.config.Test.BasePath))
+func (r *Replayer) SetupMocking(ctx context.Context, appID uint64, action MockAction, outgoingOpts models.OutgoingOptions) error {
+	if action != Start {
 		return nil
 	}
 
-	filteredMocks, unfilteredMocks, err := r.GetMocks(ctx, testSetID, afterTime, beforeTime)
+	outgoingOpts.Rules = r.config.BypassRules
+	outgoingOpts.MongoPassword = r.config.Test.MongoPassword
+	outgoingOpts.SQLDelay = time.Duration(r.config.Test.Delay)
+	outgoingOpts.FallBackOnMiss = r.config.Test.FallBackOnMiss
+	outgoingOpts.Mocking = r.config.Test.Mocking
+
+	err := r.instrumentation.MockOutgoing(ctx, appID, outgoingOpts)
 	if err != nil {
+		utils.LogError(r.logger, err, "failed to mock outgoing")
 		return err
 	}
+	return nil
+}
 
-	if action == Start {
-		outgoingOpts.Rules = r.config.BypassRules
-		outgoingOpts.MongoPassword = r.config.Test.MongoPassword
-		outgoingOpts.SQLDelay = time.Duration(r.config.Test.Delay)
-		outgoingOpts.FallBackOnMiss = r.config.Test.FallBackOnMiss
-		outgoingOpts.Mocking = r.config.Test.Mocking
+func (r *Replayer) FilterAndSetMocks(ctx context.Context, appID uint64, filtered, unfiltered []*models.Mock, afterTime, beforeTime time.Time, totalConsumedMocks map[string]models.MockState) error {
+	if !r.instrument {
+		r.logger.Debug("Keploy will not filter and set mocks when base path is provided", zap.Any("base path", r.config.Test.BasePath))
+		return nil
+	}
 
-		err = r.instrumentation.MockOutgoing(ctx, appID, outgoingOpts)
-		if err != nil {
-			utils.LogError(r.logger, err, "failed to mock outgoing")
-			return err
+	filtered = pkg.FilterTcsMocks(ctx, r.logger, filtered, afterTime, beforeTime)
+	unfiltered = pkg.FilterConfigMocks(ctx, r.logger, unfiltered, afterTime, beforeTime)
+
+	filterOutDeleted := func(in []*models.Mock) []*models.Mock {
+		out := make([]*models.Mock, 0, len(in))
+		for _, m := range in {
+			// treat empty/missing names as never consumed
+			if m == nil || m.Name == "" {
+				out = append(out, m)
+				continue
+			}
+			if k, ok := totalConsumedMocks[m.Name]; !ok || k.Usage != models.Deleted {
```

Comment:
Add a comment here that we picking both the updated mocks and all the mocks that were not consumed and can be consumed in the future.
Also try if you can add round brackets in the later condition.